File: 1912/Domain/HeadQuarter.cs
using System;
using System.Collections.Generic;

namespace _1912.Domain
{
    class HeadQuarter
    {

        private List<Commander> russianCommanders;
        private List<Commander> frenchCommanders;

        private List<Commander> Shuffle(List<Commander> toShuffle)
        {
            List<Commander> data = toShuffle;
            for (int i = data.Count - 1; i >= 1; i--)
            {
                var rand = new Random();
                int j = rand.Next(i + 1);
                // обменять значения data[j] и data[i]
                Commander temp = data[j];
                data[j] = data[i];
                data[i] = temp;
            }

            return data;
        }

        public HeadQuarter(List<Commander>ruCommanders, List<Commander> frCommanders)
        {
            russianCommanders = Shuffle(ruCommanders);
            frenchCommanders = Shuffle(frCommanders);
        }

        public Commander PickRU()
        {
            return pick(Commander.CommanderSide.RussianEmpire);
        }

        public Commander PickFR()
        {

            return pick(Commander.CommanderSide.FrenchEmpire);
        }

        public Commander Swap(Commander toSwap)
        {
            return pick(toSwap.Side, toSwap);
        }

        private Commander pick(Commander.CommanderSide side, Commander toSwap = null)
        {
            Commander toReturn = null;
            if (side == Commander.CommanderSide.RussianEmpire)
            {
                toReturn = russianCommanders[0];

                russianCommanders.Remove(toReturn);
                if (toSwap != null)
                {
                    russianCommanders.Add(toSwap);
                }
                russianCommanders = Shuffle(russianCommanders);
            }
            if (side == Commander.CommanderSide.FrenchEmpire)
            {
                toReturn = frenchCommanders[0];

                frenchCommanders.Remove(toReturn);
                if (toSwap != null)
                {
                    frenchCommanders.Add(toSwap);
                }
                frenchCommanders = Shuffle(frenchCommanders);
            }

            return toReturn;
        }

        public override string ToString()
        {
            return russianCommanders.Count.ToString() + " ru, " + frenchCommanders.Count.ToString() + " fr";
        }
    }
}

File: domain/head_quarter.py

from typing import List
from domain.commander import Commander, CommanderSide
import random

class HeadQuarter:
    
    def __init__(self, commanders):
        self.commanders = commanders.copy()
        for key in commanders.keys():
            random.shuffle(self.commanders[key])

    def pick_ru(self):
        return self._pick(CommanderSide.RUSSIAN_EMPIRE.value)

    def pick_fr(self):
        return self._pick(CommanderSide.FRENCH_EMPIRE.value)

    def swap(self, to_swap: Commander) -> Commander:
        return self._pick(to_swap.side_id, to_swap)

    def _pick(self, side_id, to_swap: Commander = None) -> Commander:
        to_return = self.commanders[side_id].pop(0)
        if to_swap != None:
            self.commanders[side_id].append(to_swap)
        
        random.shuffle(self.commanders[side_id])
        return to_return

    def __str__(self) -> str:
        return str(self.commanders[CommanderSide.RUSSIAN_EMPIRE.value]) + ' ru, ' + str(self.commanders[CommanderSide.FRENCH_EMPIRE.value]) + ' fr'
